runtime/util/hshelp.c

```diff
@@ -482,6 +485,180 @@ areSingleSlotConstantRefsIdentical(J9ROMConstantPoolItem * romCP1, U_32 index1,
 	return utfsAreIdentical(J9ROMSTRINGREF_UTF8DATA((J9ROMStringRef *) ref1), J9ROMSTRINGREF_UTF8DATA((J9ROMStringRef *) ref2));
 }
 
+/*
+ * Helper methods for areCallSiteDataMethodsEquivalent
+ */
+
+/* ROM call site data header size (16 bits of method handle address + additional argument count) */
+static const U_16 callSiteHeaderOffset = 2;
+
+static UDATA
+isMethodHandleAField(UDATA cpType) 
+{
+	return (cpType == MH_REF_GETFIELD) || (cpType == MH_REF_GETSTATIC) || (cpType == MH_REF_PUTSTATIC) || (cpType == MH_REF_PUTFIELD);
+}
+
+static U_16*
+findBSMDataAtIndex(U_16* bsmData, U_16 bsmIndex) 
+{
+	U_16 i = 0;
+	for (i = 0; i < bsmIndex; i++) {
+		/* increment by size of bsm data plus header */
+		bsmData += (bsmData[1] + callSiteHeaderOffset);
+	}
+	return bsmData;
+}
+
+static U_32
+iterateToNextArgument(U_32 sigIndex, U_32 sigLength, U_8* sigData) 
+{
+	if (sigIndex >= sigLength) return sigIndex;
+
+	/* check for object */
+	if (sigData[sigIndex] == 'L') {
+		while((sigIndex < sigLength) && (sigData[sigIndex] != ';')) {
+			sigIndex += 1;
+		}
+	}
+	/* for an object this will move past the ;, for a primitive this will move past the argument */
+	sigIndex += 1;
+	return sigIndex;
+}
+
+static UDATA
+areCallSiteDataMethodsEquivalent(J9ROMClass* romClass1, UDATA callSiteIndex1, J9ROMClass* romClass2, UDATA callSiteIndex2)
+{
+	/**
+	 * Structure of CallSiteData structure in ROM class (also see romclasswalk.c)
+	 * NAS:
+	 * 		callSiteCount SRP to J9ROMNameAndSignature structures describing the resolved callsite. Note: SRP is 32 bits
+	 * bsm indices:
+	 * 		callSiteCount 16 bit indices into bsmData. for example 0 is the first bsm, 1 is the second etc.
+	 * bsm data:
+	 * 		(header) The first 16 bits contains the index of the bsm's MethodHandle in the constant pool.
+	 * 		(header) The second 16 bits contains the number of bsm arguments in addition to the required three (MethodHandles.Lookup, String, MethodType).
+	 * 		The additional BSM arguments, these are constant pool indices.
+	 */
```

Comment:
I think it would be good to move the documentation up above all of the code which manipulates the data (and perhaps be a bit more explicit (i.e. describe it more like C structures).

Reply:
The new doc is great - the tabbing appears inconsistent in the github view, but we haven't come up with a good solution for this - if if looks good in Eclipse, I'm fine.

Reply:
@DanHeidinga Any formatting guidance here?